Named programs

tsyesika opened the pull request:
This introduces named programs as outlined in https://github.com/snabbco/snabb/issues/1033. The two main functions are `claim_name` which is suppose to be used by the programs to claim a unique name for themselves and `enumerate_named_programs` which should list all the programs by their name and PID.

I've also incorporated code into the cleanup done by the supervisor. My one concern for this is it changes `/var/run/snabb` from purely being a list of PID's with their shared memory inside, it now adds a "by-name" folder which has the symlinks to the PIDs. For example in the top program I already have to filter out the "by-name" directory out so it's not listed. I'm concerned other programs are written based on that assumption.

A possible consideration would be to instead change the directory from `/var/run/snabb` to `/var/run/snabb/by-pid/`. This would allow for a single directory to continue being just full of PID's with their shared memory.

Commit:
Add named programs

This adds the functionality for programs to claim names and other processes
to enumerate processes by their name to retrive their PIDs. These two functions
are in core.apps so will be exposed to programs through "engine"

tsyesika committed:
Remove unintentional addition of claim_name to example_relay

I accidently left a claim_name from when I was debugging, this commit
removes that. Oops!

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,65 @@ function configure (new_config)
+      local pid = tonumber(string.sub(piddir, ps+1, pe))
+      if S.kill(pid, 0) then
+         -- It's a running process, we should display an error message.
+         error("Name has been claimed by a currently running process: "..namedirqualified)
```
Is the kill necessary?  I guess we could grovel in /proc.  One question: are you sure that the process with this PID is the one that made the name dir?  Does the supervisor process guarantee that?  If not we could write the process start time into the per-pid directory or something.  Maybe this is not important.

tsyesika replied:
Uh. You know, I'm not sure of that... Right now as the code stands we I think it is the correct PID but with luke's multiprocess changes maybe not. A way around this could be to move the `claim_name` call into `engine.main` this should fix that just have name a parameter when you call it, e.g. `engine.main(options, name)`.

WDYT?

wingo replied:
My concern is a stale by-name directory for a dead PID; how does moving around the place where we claim the name affect that?

Also we should be sure to have the supervisor clean up the by-name directory as well I would guess.

tsyesika replied:
Okay so.

I've understood the issue with stale PIDs. There is some mitigation against any problems they would cause.
1. `claim_name` checks for stale PID's using kill 0 and if it is dead, it'll unlink it.
2. `enumerate_named_programs` also uses kill 0 on each named program it discovers and only enumerates those which are still alive.
3. I've added to the shutdown function in which is run by the supervisor to unlink named programs which are being shutdown. 

This should mean that directories are cleaned up, it also should mean if something happens and they're not cleaned up programs can still claim the name as it will be found at that stage and unlinked so it can be re-used and finally other programs enumerating these would only get named programs which are still active at the time of calling `enumerate_named_programs`

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,65 @@ function configure (new_config)
+-- it's still alive. Any dead program or program without a name is not listed.
+function enumerate_named_programs()
+   local progs = {}
+   local dirs = S.util.dirtable(named_program_root, true)
```
Is this the only instance of S.util in this file?  If so probably we shouldn't use it.  Does `core.lib.files_in_directory` work for you?  (Though, ugh, shelling out to ls!!!!!)

tsyesika replied:
It is but i also use syscalls to do a kill 0 to see if the process is still around, as you said above I suppose i could poke around in `/proc`, kill just seems... more straight forward but either way works.

wingo replied:
Use of syscalls is fine, I think use of syscall helpers is probably not.

wingo commented on src/core/main.lua:
```diff
@@ -143,6 +144,14 @@ end
    if not _G.developer_debug and not lib.getenv("SNABB_SHM_KEEP") then
       shm.unlink("/"..pid)
+
+      -- Look through the named apps and unlink any which are for this process.
```
WDYT about maintaining a back-link from the process to its name?

wingo replied:
Yes I think a backlink is the right thing here, please add one so that we can avoid this loop and the one in top.lua.

wingo reviewed (commented):
Looking good to me :)  Some questions.

tsyesika committed:
Replace usage of syscall.util and fix cleanup code

This replaces the use of syscall.util so that core.app no longer uses
that. It also fixes the supervisor which wasn't cleaning up thhe named
app symlinks - it now does.

wingo reviewed (changes requested):
Please avoid looking through the whole /run/snabb for normal operations.  Also please submit upstream for feedback in parallel; see SnabbCo#1033.

tsyesika committed:
Add backlink for named programs

xray7224 commented:
I have pushed an update which adds a backlink, this create's a symlink in the program's PID directory which links to the name program symlink in "by-name". I've also put in a check to ensure a program can't claim multiple names.

wingo commented on src/program/top/top.lua:
```diff
@@ -32,7 +32,9 @@ function run (args)
 function select_snabb_instance (pid)
-   local instances = shm.children("/")
+   local instances = {}
+   -- For named programs there is a "by-name" directory which isn't a process, remove it.
```
Would you mind doing a little refactor while you're at it -- please add something like

``` lua
local function compute_snabb_instances()
   --- Compute the set of Snabb processes, excluding this one.
   local pids = {}
   local my_pid = S.getpid()
   for _,name in ipairs(shm.children('/')) do
      local pid = tonumber(name) -- Could fail for the "by-name" directory.
      if pid and pid ~= my_pid then table.insert(pids, pid) end
   end
   return instances
end
```

and refactor the select_snabb_instance function to be simpler.

wingo commented on src/core/main.lua:
```diff
@@ -142,6 +143,14 @@ end
 function shutdown (pid)
    if not _G.developer_debug and not lib.getenv("SNABB_SHM_KEEP") then
+      -- Check if there is a backlink to the named app, if so cleanup that.
+      local backlink = shm.root.."/"..pid.."/name-backref"
```
Can we just call this one "name" ?

wingo commented on src/core/app.lua:
```diff
@@ -526,4 +612,19 @@ function selftest ()
+   -- Ensure to claim two names fails
+   assert(not pcall(claim_name, basename.."2"))
+
+   -- Lets check if they can be enumerated.
```
"Let's".

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
 
+-- Claims a name for a program so it can be identified by name by other processes.
+--
+-- The name given to the function must be unique, if a name has been used before by
```
Fix comma-splice please by replacing comma with either semicolon or period.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+--
+-- The name given to the function must be unique, if a name has been used before by
+-- an active process the function will error displaying an appropriate error message.
+-- Any other process can enumerate programs by their name and PID using enumerate_named_programs.
```
Please check 80 column limit.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+   function name_from_path(dir)
+      -- Extracts last file/folder name from path
+      -- e.g. "/var/run/snabb/hello" -> "hello"
+      local ds, de = string.find(dir, "/[^/]*$")
```
Use core.lib.basename please

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+   -- Verify that the by-name directory exists.
+   shm.mkdir(namedir)
+
+   -- Verify that we've not already claimed a name
```
Let's keep a global variable instead of doing the readlink, e.g. instead do `assert(not claimed_name)`.  You can probably elide the error message too since this is a simple coding error.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+      error("Program already claimed name '"..fname.."', cannot claim name '"..name.."'")
+   end
+
+   -- Check if a symlink to a named program with the same name exists, if not, unlink it.
```
Let's elide this section (lines 161-173).  Even with this section there's a TOCTTOU bug, but one which is robustly handled by the S.symlink call later, so this check is needless.  In theory we never get into this situation anyway because of the supervisor process.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+   
+   -- Create the new symlink.
+   local rtn, err = S.symlink(piddir, namedirqualified)
+   if rtn == -1 then
```
S.symlink returns nil on error, so best to replace with an assert around the S.symlink -- in an error case that will end up calling assert(nil, err) (because of the way Lua's multiple-value support works out) and then we get an appropriate error message.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+      error("Error creating program name symlink in: "..namedirqualified)
+   end
+
+   -- Create a backlink so to the symlink so we can easily cleanup
```
Likewise.

wingo commented on src/core/app.lua:
```diff
@@ -124,6 +128,88 @@ function configure (new_config)
+   for _, program in pairs(dirs) do
+      local fq = named_program_root .. "/" .. program
+      local piddir = S.readlink(fq)
+      local s, e = string.find(piddir, "/[^/]*$")
```
Use lib.basename here and below.

wingo reviewed (changes requested):
Hi!  Great patch, though I have some nits.  I think we can tighten this up a bit; wdyt?

tsyesika committed:
Refactor top.select_snabb_instance and fix small issues in claim_name

This fixes a number of issues, some of which are cosmetic. The change
refactors some of the "top" program's code for PID locating and
selection. This change also cleans up a number of checks in claim_name
into asserts.

The big change is removal of the top level import of "core.app" into
main. There doesn't seem to be usage of this paticular import in main
and it causes stale data in "/var/run/snabb" as it creates "engine" shm
directories for the supervisor.

tsyesika commented on src/core/main.lua:
```diff
@@ -11,7 +11,6 @@ local ffi = require("ffi")
 local zone = require("jit.zone")
 local lib = require("core.lib")
 local shm = require("core.shm")
-local app = require("core.app")
```
I want to draw attention to this change specifically.

I was noticing a bug in the code where upon running a program such as "top" with no arguments and no other processes running it'd still connect to a process and display the counter information. It turned out because we were including app in main.lua and app at the top level sets up a counter directory for the engine the supervisor process actually had their own PID directory which was being created a not cleaned up. This also caused an issue that the process top was connecting to was it's supervisor process which was a little nonsensical.

I looked through the code and verified nothing seems to be using the import, removing it and running all the tests worked fine and also trying to run a few snabb programs seems to work too. Not sure if this was left in there from a time when it was used.

I just want to check there is no other magic that it's doing and it's safe to remove.

tsyesika replied:
**Update**: when trying to report this upstream I realized it wasn't occuring, on further examination I think I added this import and I can't find it in the `lwaftr` branch or upstream master. Feel a little bit silly now!

Anyway. I think it's fine to remove it as I think I must have added it, causing the bug in the first place.

xray7224 commented:
I think everything you've mentioned has been addressed. Can you take another look @wingo?

wingo commented on src/core/main.lua:
```diff
@@ -142,6 +142,14 @@ end
    if not _G.developer_debug and not lib.getenv("SNABB_SHM_KEEP") then
+      -- Check if there is a backlink to the named app, if so cleanup that.
+      local backlink = shm.root.."/"..pid.."/name"
+      if S.lstat(backlink) then
```
Would you mind removing this if?  Instead if the readlink on the next line fails, you will get a nil name_link, and in that case you replace line 149 with if name_link then S.unlink(name_link) end, and you ignore errors in any case.

wingo commented:
Changes lgtm however the bot has found some errors related to this patch, can you have a look please.

tsyesika committed:
Fixes error in selftest for core.app

The assert was checking the opposite of what it was
suppose to, failing when things were correct. This
fixes it to ensure it is checking that program has NOT
got a name assiged.

wingo commented on src/core/main.lua:
```diff
@@ -142,13 +142,11 @@ end
-         S.unlink(backlink)
-      end
+      local name_link = S.readlink(backlink)
+      S.unlink(name_link)
```
Does this fail silently if name_link is nil?  That would be a bit surprising.

tsyesika replied:
It returns nil, "Bad Address" which is silently enough for us :)

wingo commented:
The bot error seems to be a timeout.  Would you mind making some minor cleanup and pushing another patch on this PR just to force the bot to run again?  Thanks :)

tsyesika committed:
Minor cleanup of comment to re-trigger CI bot.
